lambda/shared/src/main/scala/feral/lambda/tracing/TracedLambda.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.tracing
+
+import cats._
+import cats.data._
+import cats.effect.std.Random
+import cats.effect.{Trace => _, _}
+import natchez._
+import feral.lambda._
+import natchez.noop.NoopSpan
+import natchez.xray.{XRay, XRayEnvironment}
+
+abstract class TracedLambda[F[_]: MonadCancelThrow, G[_], Event] {
+  def extractKernel(event: Event, context: Context[F]): F[Kernel]
+
+  def runTraced[A](span: Span[F])(ga: G[A]): F[A]
+
+  def apply[Result](entryPoint: EntryPoint[F], fk: F ~> G)(
+      lambda: Lambda[G, Event, Result]): Lambda[F, Event, Result] =
+    (event, context) =>
+      Resource
+        .eval(extractKernel(event, context))
+        .flatMap { kernel => entryPoint.continueOrElseRoot(context.functionName, kernel) }
+        .use { span => runTraced(span)(lambda(event, context.mapK(fk))) }
+}
+
+abstract class KleisliTracedLambda[F[_]: MonadCancelThrow, Event]
+    extends TracedLambda[F, Kleisli[F, Span[F], *], Event] {
+  override def runTraced[A](span: Span[F])(ga: Kleisli[F, Span[F], A]): F[A] = ga(span)
+}
+
+sealed abstract class TracedIO[A] {
+  def run(span: Span[IO]): IO[A]
+}
+
+object TracedIO {
+  def apply[A](f: Trace[IO] => IO[A]): TracedIO[A] =
+    new TracedIO[A] {
+      def run(span: Span[IO]): IO[A] = Trace.ioTrace(span).flatMap(f)
+    }
+}
+
+abstract class TracedIOLambda[Event] extends TracedLambda[IO, TracedIO, Event] {
+  override def runTraced[A](span: Span[IO])(ga: TracedIO[A]): IO[A] = ga.run(span)
+}
+
+class XRayKleisliTracedLambda[F[_]: Sync, Event] extends KleisliTracedLambda[F, Event] {
+  override def extractKernel(event: Event, context: Context[F]): F[Kernel] =
+    XRayEnvironment[F].kernelFromEnvironment
+}
+
+object XRayKleisliTracedLambda {
+  def apply[F[_]: Async, Event, Result](
+      installer: Resource[
+        Kleisli[F, Span[F], *],
+        Lambda[Kleisli[F, Span[F], *], Event, Result]]): Resource[F, Lambda[F, Event, Result]] =
```

Comment:
Yeah, I'm not crazy about this 😆 I need to think about this more, but I don't think the lambda middlewares should need to know anything about resources. Any resources they need (such as an http client or natchez entrypoint, they should ask for in the form of arguments. The user can inject these as they build their resource stack with the stuff they want.

An alternative encoding, if we want to prescribe a particular resource, could be that the middleware _itself_ is contained inside a resource, like:
```scala
Resource[F, Lambda[G, Event, Result] => Lambda[F, Event, Result]]
```

Reply:
I started with only the `Lambda[Kleisli[F, Span[F], *], Event, Result] => Resource[F, Lambda[F, Event, Result]]` variant, but I found that everywhere I wanted to use it, I was having to write the body of this one. So this is really just an alternate convenience constructor that includes the `installer.mapK(Kleisli.applyK(new NoopSpan[F]))` that the user would need to include otherwise.

Reply:
Given that you need trace the resource, I understand why this helps. But I'm still confused why that is. I haven't really played with it yet so I'm probably missing something.

Reply:
[Here's an example](https://github.com/Dwolla/rabbitmq-topology-backup/blob/1b885b524691ba08940c7e4ff74bb5481b583ac4/src/main/scala/com/dwolla/rabbitmq/topology/LambdaHandler.scala#L28-L51):

My resource needs a `Trace[F]` constraint because it creates resources that themselves require `Trace[F]`. 

```scala
def handler[F[_] : Async : Trace]: Resource[F, lambda.Lambda[F, RabbitMQConfig, Unit]] = …
```

With the convenience method, I can just pass that `handler[Kleisli[IO, Span[IO], *]]` to `XRayTracedLambda`:

```scala
override def run: Resource[IO, lambda.Lambda[IO, RabbitMQConfig, Unit]] =
  XRayTracedLambda(handler[Kleisli[IO, Span[IO], *]])
```

Reply:
I see, thanks for the example. Makes sense, but I don't like it 😆 I want to think about this more.

Btw, just an aside: for lambdas with `Unit` result type, I've been wondering if actually we should use `Nothing` as the type parameter. The only valid instance of `Option[Nothing]` is `None` which is exactly what we want :)

Reply:
I took a more careful look at your example. I don't completely follow the `KmsAlg.resource[F].map(_.withTracing)` part so IDK if my comments apply to that, but I think they certainly apply to ember client.

I'm fairly certain that we should be distinguishing between `F` and `G` here, like so: `Resource[F, Client[G]]`. Or more precisely, you should build the `Client[F]` in untraced `F` and `mapK` it to a `Client[G]` in traced `G`, for interop with the NatchezMiddleware. It seems that `Client` lacks a `mapK` method, but I'm not sure why, hopefully it's just an oversight.

As I see it, the problem is if you build the client in `F`, and `F` is traced, that means that the root span has to exist at the time when you _create_ the client. But that means all requests from that client are part of the same span! This certainly doesn't make sense, if the client is re-used for multiple events that themselves start their own spans.

Reply:
Aha, on `Client` it is called `translate` :)
https://http4s.org/v1.0/api/org/http4s/client/client#translate[G[_]](fk:F~%3EG)(gK:G~%3EF)(implicitevidence$1:cats.effect.Async[G],implicitF:cats.effect.MonadCancel[F,Throwable]):org.http4s.client.Client[G]

Reply:
> Btw, just an aside: for lambdas with Unit result type, I've been wondering if actually we should use Nothing as the type parameter. The only valid instance of Option[Nothing] is None which is exactly what we want :)

Yeah, I've thought a little bit about that too. It seems like it makes the type inference worse when `Nothing` is the return type, for some reason. e.g.

```scala
def handler[F[_] : Async, G[_] : Async : Trace](implicit LT: LiftTrace[F, G]): Resource[F, lambda.Lambda[G, Event, Nothing]] = ???

override def run: Resource[IO, lambda.Lambda[IO, Event, Nothing]] =
  handler[IO, Kleisli[IO, Span[IO], *]].flatMap(XRayTracedLambda.usingEnvironment[IO, Kleisli[IO, Span[IO], *], Event, Nothing](_))
```

On 2.13.7, it's unable to infer that `Result` should be `Nothing` for the `XRayTracedLambda.usingEnvironment` call there. (When `Result` is `Unit`, I was able to just `.flatMap(XRayTracedLambda.usingEnvironment(_))`.)

If we want to encourage using `Nothing`, I think we'd also want to include something like this, since `IOLambda` demands an `Encoder[Result]`:

```scala
@nowarn("msg=dead code following this construct")
implicit val nothingEncoder: Encoder[Nothing] = (_: Nothing) => Json.Null
```

Reply:
Hmm, seems like an uphill battle. No harm sticking to unit :) I forgot about the encoder thing, I think I remember trying this before and immediately running into that.

Reply:
Oh, maybe the type inference thing is why `fs2` has `INothing`.

Reply:
Oh, yeah, `INothing` infers fine. Nice!

Reply:
> I don't completely follow the KmsAlg.resource[F].map(_.withTracing) part

The `.withTracing` enhancement method uses cats-tagless to auto-instrument an algebra using the class and method name as the span name. This lets me take an algebra and wrap its methods in tracing with very little boilerplate, because almost everything is auto-derived at compile time.

When we separate `F[_]` and `G[_] : Trace` (given `LiftTrace[F, G]`), it ends up looking like

```scala
KmsAlg.resource[F].map(_.mapK(LT.liftK).withTracing)
```

which works but I don't love—it feels like separating the effects introduces more boilerplate for no gain when using Kleisli trace. It'd be nice to know for sure whether the separation is required for the IOLocal trace.

Reply:
Woah, that is very cool!!

> which works but I don't love

Yeah, this `F` and `G` stuff is technically sound but a real bummer in practice. I had a chat with @ChristopherDavenport in Discord, and he strongly encourages using `Span[F]` directly instead of `Trace` precisely to avoid these complexities.
https://discord.com/channels/632277896739946517/632286375311573032/910352732353863770

> It'd be nice to know for sure whether the separation is required for the IOLocal trace.

Are you referring to my half-baked thought from above?

> As I see it, the problem is if you build the client in F, and F is traced, that means that the root span has to exist at the time when you create the client. But that means all requests from that client are part of the same span! This certainly doesn't make sense, if the client is re-used for multiple events that themselves start their own spans.

The answer for `IO` is ... it depends 😆 

To be more specific: `IO` is special, because (unlike Kleisli) it can function as both the untraced and traced effect :). So, if you _know_ your concrete type is `IO`, then you can build all this stuff without the terrible `mapK`s and `translate`s because everything is `IO` and life is good.

_However_, if you try to build everything in generic `F` you get into trouble real fast. E.g.
```scala
def mkTracedClient[F: Async: Trace]: Resource[F, Client[F]] =
  clientBuilderThing.build[F].map(natchezTracingThing[F])

// Now let's use it with IO :)
// We need a Trace[IO]
Trace.ioTrace(myRootSpan) // Boom! You just installed your root span, game over.
  .flatMap { implicit trace =>
    mkTracedClient[IO].use {
      // etc.
    }
  }
```

Basically, the problem is you only have one chance to insert a root span, and that's at the time you create the `Trace[IO]`. So if you do it too early, then you're stuck with whatever root you had at that point.

Apologies if I'm being dumb, but let me know if that makes sense.